### ISSUE TITLE
[RFC]: inductor pass and npugraph ex pass normalization

### ISSUE THREAD

**Issue by wxsIcey** [vllm-project/vllm-ascend#6347]
### Motivation.

Currently, the inductor pass and npugraph ex pass coexist, and are stored in the `compilation/passes` and `compilation/npugraph_ex_passes` directories, respectively. These two passes are supported by two different backends. In the short term, we will switch to npugraph ex to support custom passes, but we need to have two sets of passes coexisting. This is because:

**1. inductor is the default backend supported by vllm**

Inductor is the default backend of vLLM. Since we do not support codegen of inductor, we currently only use the pass capability. Once the codegen capability is available, we will reconsider using the inductor backend.

**2. fully utilize the additional capabilities of npugraph ex**

npugraph ex extends the aclgraph function based on `torch.compile`. In addition to the capability of customizing passes, npugraph ex also has capabilities such as redundant kernel optimization and kernel performance optimization.

### Proposed Change.

I propose to design a set of common passes for the two backends.

- **Remove the `compilation/npugraph_ex_passes` directory**

   If two nearly identical passes exist in two different directories, developers need to add these passes to both directories when adding new passes.

### Design Overview.
I create a base class named `BasePattern`. The `register` method of this class implements pattern registration for two different backends. All passes inherit from this base class and implement their own `example_inputs`, `pattern`, and `replacement` methods.

**BasePattern：**
```python
class BasePattern(ABC):
    def __init__(self, vllm_config: VllmConfig):
        self.vllm_config = vllm_config
        self.dtype = vllm_config.model_config.dtype
        
    @abstractmethod
    def get_example_inputs(self) -> List[torch.Tensor]:
        pass
    
    @abstractmethod
    def get_pattern(self) -> Callable:
        pass
    
    @abstractmethod
    def get_replacement(self) -> Callable:
        pass
    
    def get_extra_check(self) -> Optional[Callable]:
        return None
    
    def register(self, pm_pass: PatternMatcherPass) -> None:
        pattern_fn = self.get_pattern()
        replacement_fn = self.get_replacement()
        example_inputs = self.get_example_inputs()
        extra_check = self.get_extra_check()
        
        pm.register_replacement(
            pattern_fn, replacement_fn, example_inputs, 
            pm.fwd_only, pm_pass
        )
        
        torchair.register_replacement(
            search_fn=pattern_fn,
            replace_fn=replacement_fn,
            example_inputs=example_inputs,
            extra_check=extra_check or extra_stream_scope_check,
        )
```
**Specific implementation of pass：**
```python
class AddRMSNormQuantPattern(BasePattern):
    def __init__(self, vllm_config: VllmConfig, eps: float = 1e-6):
        super().__init__(vllm_config)
        self.eps = eps

    def get_example_inputs(self):
        """
        Generate example inputs for the AddRMSNormQuant fusion pattern.
        """
        rms_norm_input = torch.randn(2, 4, device="npu", dtype=self.dtype)
        residual = torch.randn(2, 4, device="npu", dtype=self.dtype)
        rms_norm_weight = torch.randn(4, device="npu", dtype=self.dtype)
        scale = torch.ones(4, device="npu", dtype=self.dtype)
        scale_reciprocal = torch.ones(4, device="npu", dtype=self.dtype)
        offset = torch.zeros(4, device="npu", dtype=self.dtype)
        return [rms_norm_input, residual, rms_norm_weight, scale, scale_reciprocal, offset]

    def get_pattern(self):
        def pattern(
            rms_norm_input: torch.Tensor,
            residual: torch.Tensor,
            rms_norm_weight: torch.Tensor,
            scale: torch.Tensor,
            scale_reciprocal: torch.Tensor,
            offset: torch.Tensor,
        ):
            """
            Pattern for AddRMSNormQuant fusion.
            """
            output = torch.ops.npu.npu_add_rms_norm(rms_norm_input, residual, rms_norm_weight, self.eps)
            out0 = output[0]
            out1 = output[2]
            quantized_output = torch.ops.vllm.quantize(out0, scale, scale_reciprocal, offset)
            return quantized_output, out1
        return pattern

    def get_replacement(self):
        def replacement(
            rms_norm_input: torch.Tensor,
            residual: torch.Tensor,
            rms_norm_weight: torch.Tensor,
            scale: torch.Tensor,
            scale_reciprocal: torch.Tensor,
            offset: torch.Tensor,
        ):
            """
            Replacement for the AddRMSNormQuant fusion.
            """
            output = torch.ops.npu.npu_add_rms_norm_quant(
                rms_norm_input, residual, rms_norm_weight, scale, offset, epsilon=self.eps
            )
            quantized_output = output[0]
            out1 = output[2]
            return quantized_output, out1
        return replacement 
```

### Feedback Period.

one week

### CC List.

@wangxiyuan @Yikun @whx-sjtu @panchao-hub @ChenCangtao @ForBetterCodeNine 

### Any Other Things.

- It should be noted that, according to this RFC, when contributors add passes, they should ensure that these passes tests on both backends and meet the performance requirements.
- I'm unsure which passes are currently integrated into npugraph ex, this needs to be considered when adding passes later.
